Documentation/proposal-runconfig_inside_osv.md

```diff
@@ -1,55 +1,286 @@
+Currently, Capstan runtime support makes use of `$CWD/meta/run.yaml` for composing (COMPOSE) and
+running (RUN) an OSv image. Although this is fine for the composing, it gets tricky for running
+because `capstan run <IMAGE>` generally does not require user to be inside package root directory.
+
+Furthermore, you currently need Capstan if you want to pick a desired named configuration to RUN
+pre-composed OSv image with. That's because capstan is able to:
+* understand your meta/run.yaml
+* generate run command from it
+* set run command to the existing OSv image
+
+Other infrastructures (OpenStack, Amazon) are unable to perform first and second bullet, but can do
+the third one.
+
+### Executive summary
+This proposal is about what actions will Capstan perform on COMPOSE time based on user-friendly
+input file that is provided in `$CWD/meta/run.yaml`. The format of this file was decided [here](https://github.com/mikelangelo-project/capstan/issues/3#issuecomment-245210949) and is not to be changed much with this proposal
+(values starting with $ will be parsed as placeholder name, that's the only difference).
+
+The idea is that Capstan would prepare everything in COMPOSE time so that the first and second bullet
+(see Problem description section) would not be needed at the RUN time. This way OpenStack would be
+able to make use of named runtimes by not even having Capstan installed.
 
 I propose the following solution:
 
+### User aspect
+OSv packaging with runtime support ends up with four user roles:
+
+##### Application user
+Downloads a pre-composed image and RUN it without modifications. She uses EXTERNAL packages and
+EXTERNAL application. She does not perform COMPOSE step. She must be provided with this information:
+ * what configuration names are available
+ * what parameters are needed for each configuration name
+
+Example: Hdfs. User downloads it and runs it with customized parameters.
+
+##### Runtime user
+Downloads pre-prepared Capstan packages. She uses EXTERNAL packages and INTERNAL application. She
+needs to provide the `meta/run.yaml` file. There she needs to pick one of the supported runtimes that suites her best. Then she needs to either provide runtime-specific values either define placeholder
+names for them (explained in following sections). She creates arbitrary number of configurations and
+names them.
+
+To successfully perform the COMPOSE step, she must be provided with this information:
+ * what runtimes are available
+ * what runtime-specific attributes are supported
+
+Once she successfully composes the image, she takes the role of Application user.
+
+Example: NodeJS. User uses 'node' runtime (which downloads pre-prepared node package), but provides
+her own NodeJS JavaScript application code.
+
+##### Runtime developer
+Extends Capstan tool by implementing runtime.Runtime interface (Golang programmer). She must be
+provided with this information:
+ * what packages are available
+ * how to use package
```

Comment:
Not sure I understand why runtime dev needs this information? My assumption is that this role also provides the main package that must be included by the runtime, right?